examples/network_traffic_prediction/nr_models.py

```diff
@@ -0,0 +1,702 @@
+            test_label = None
+        ntype = list(val_label.keys())[0]
+
+        val_pred = val_pred[ntype].to(self.device)
+        val_label = val_label[ntype].to(self.device)
+        if test_pred is not None:
+            test_pred = test_pred[ntype].to(self.device)
+            test_label = test_label[ntype].to(self.device)
+        
+        val_score, test_score = self.evaluator.evaluate(val_pred, test_pred,
+                                                        val_label, test_label, total_steps)
+        if gs.get_rank() == 0:
+            self.log_print_metrics(val_score=val_score,
+                                    test_score=test_score,
+                                    dur_eval=time.time() - teval,
+                                    total_steps=total_steps)
+        return val_score
+
+class NodePredictionInferrer4TS(GSgnnNodePredictionInferrer):
+    """ Simply overwrite GSgnnNodePredictionInferrer's ``infer()`` method, using mini-batch
+    inference only.
+    
+    For inference, need to use autoregresion method because we can not get real value as time
+    window input, need to use predicted value to predict the next ones.
+    """
+    def infer(self, loader, save_embed_path, save_prediction_path=None,
+              use_mini_batch_infer=False,
+              node_id_mapping_file=None,
+              return_proba=True,
+              save_embed_format="pytorch",
+              init_step=0,
+              infer_steps=14):
+        """
+        New arguments
+        --------------
+        init_step: int
+            The starting point for prediction. That is, the saved model will need to use historical
+            data from the init_step to (init_step + model.window_size) to predict the value in the
+            (init_step + model.window_size) + 1 step.
+        infer_steps: int
+            Steps to be predicted. That is, the save model will predict from init_step
+            (init_step + model.window_size), and run infer_steps. When sliding one window ahead, use
+            predicted value as input, and predict the next ones.
+        """
+
+        # ====== Not do evaluation in the auto-agressive inference ====== #
+
+        # ====== Not save embedding in this customized inferrer ========= #
+        assert save_embed_path is None, \
+            "Unable to save the node embeddings when using mini batch inference." \
+            "It is not guaranteed that mini-batch prediction will cover all the nodes."
+
+        self._model.eval()
+
+        # get the graph
+        g = loader.data.g
+        # get the model metadata
+        window_size = self._model.window_size
+
+        # This customized inferre only handle the first ntypes for inference.
+        infer_ntype = list(loader.target_nidx.keys())[0]
+
+        # get all demands features
+        inventory_nfeats = g.nodes[infer_ntype].data['inventory_amounts']
+        th_inventory_nfeats = inventory_nfeats[th.arange(inventory_nfeats.shape[0])]
```

Comment:
what does th_ mean?

Reply:
means torch. This variable serves as a temporary torch tensor used to replace existing node features.